### ISSUE TITLE
Repeatability of games with assigned seed

### ISSUE THREAD

**Issue by jkiliani** [gcp/leela-zero#361]
Over the last few hours, I got suspicious when in the match be8b358e VS fe7a6c28, I got 9 straight wins of white, many of them very short games where black passed early:
[LeelaZ_log.txt](https://github.com/gcp/leela-zero/files/1551369/LeelaZ_log.txt)

I halted autogtp and attempted to replicate some of these games with Sabaki, using the exact same command line parameters (including seed) except for -q (since I was interested in diagnosing). To my surprise, game number 6 (the one only 4 moves long) generated by autogtp turned out not to be repeatable. Even more surprising, when I did repeat runs with Sabaki, using the same players and identical seeds, every game was different.

Can some other people confirm that this is not an issue only with my machine? Are autogtp-generated (match) games really repeatable?

**Comment by Yakago** [comment]
Somehow I find it disturbing that so many of these end with early passing

**Comment by zhu-jz** [comment]
They are repeatable as long as using the same seeds

**Comment by jkiliani** [comment]
That's not an answer. They're obviously supposed to be repeatable, but for me, they aren't. Did you actually test that they are repeatable for your case?

**Comment by jkiliani** [comment]
@Yakago I found that disturbing as well, which is the reason I tried to diagnose how the search had decided on this early black pass in the first place. But without repeatability, I don't see how I can investigate...

**Comment by barrtgt** [comment]
Aren't there race conditions that introduce randomness existent in search anyway?

**Comment by jkiliani** [comment]
I don't know, that's why I'm asking. But there's really no point in the option to specify a seed if that seed doesn't ensure repeatability.

**Comment by jkiliani** [comment]
If @roy7 reads this, I propose a new "job": Assign a number of clients (~100) to produce a match between two arbitrary networks, using identical seeds for each match. Then compare what the clients report back.

**Comment by NTT123** [comment]
@jkiliani Can you try with small playouts, e.g., `-p 10` ?
I doubt that the cause is about floating-point-related problem. So, my guess is that with small playouts, only at later moves, say, move 100, you will see the difference.

**Comment by jkiliani** [comment]
@NTT123 I just tested with 10 playouts. The games looks more similar compared to 1600 playouts, but they still diverge within the first 10 moves.

Can you confirm repeatability on your machine?

**Comment by NTT123** [comment]
@jkiliani, yes, I confirmed that.

 I run  `./leelaz -w 5bcb392e581915631ab7dd43c08ce7aae9ab51e2ff6c81793e54e69d55b6073f  -t 1 -r 30 -m 1 -p 5 -d  --noponder  -s 12`

then, I use `auto` command

and repeatedly having `final_sore` command outputs `=B+56.5`

**Comment by jkiliani** [comment]
@NTT123 Well, that proves that repeatability works on your machine for a game played internally by a single instance of LeelaZ on a single network, with the parameters you specified. Question is: Does that prove that using LeelaZ on autogtp or Sabaki with two opposing instances also generates repeatable results?

I purposely kept as closely to the parameters used by autogtp as possible when testing in Sabaki. If repeatability works for you, does that mean you can recreate a match-type autogtp game afterwards?

**Comment by NTT123** [comment]
@jkiliani, as I know, `auto` command is the same as `genmove b/w` alternating which are used in autogtp. So, I'm quite sure (but not tested) that autogtp will have the same repeatability as `auto` does. 

btw, I searched all autogtp directory, it seems having no option for setting random seed there.

**Comment by marcocalignano** [comment]
@NTT123  The random seed is passed directly to the leelaz process when it start, but it is an option implemented in the /next branch.

**Comment by jkiliani** [comment]
I know from own experience that keeping code deterministic (was a scientific model then) is a programming minefield, where lots of things can go wrong. If it's not tested under realistic conditions, and with multiple clients/setups, something usually goes wrong, or even the deterministic behaviour is lost again after you already had it once, with a minor code change you didn't even consider might have that effect.

All I'm saying here is: The deterministic behaviour evidently doesn't work for me. I hope that other programmers here will do some testing on their own machines, under realistic conditions (i.e., really recreate a /next autogtp game from the assigned seed)

**Comment by killerducky** [comment]
Two issues here:

1) Repeatability: Random seeds are in v0.9, but there was a bug (see #300). @jkiliani if you're able, try compiling the latest from next. Otherwise you need to wait for the next release.

2) The quicker games and early passing. This was also reported in #359. Are you using an AMD GPU?

**Comment by jkiliani** [comment]
@killerducky I am compiling the latest from next. I'm using an Intel Iris 6100 on a notebook.

**Comment by killerducky** [comment]
Ok well it seems there is still a bug with repeatability then. I'll test it on my machine tonight after work, but it's probably hardware specific since at least a few people say it is working.

**Comment by jkiliani** [comment]
I agree, which is why I proposed that the devs should consider testing the repeatability on the server setup, by assigning seed-identical assignments to a considerable number of clients. If most clients report the correct game, that would prove the impact of the issue is small. If many report incorrect ones, however...

**Comment by Ttl** [comment]
The games are not reproducible between computers using different hardware. However the same computer should play the same game when using the same seed. See #300.

**Comment by jkiliani** [comment]
@Ttl Thanks, I didn't know that. Hope that won't become a roadblock about the plan to make LeelaZ Boinc-compatible...

**Comment by killerducky** [comment]
@jkiliani two more thoughts:

1) Just to double check. I can in the logfile you have autogtp v10. But can you double check that you recompiled leelaz in the src directory and copied the new binary into autogtp dir? (I use a symbolic link to reduce the chance of errors).

2) When you run in sabaki, did you connect two engines, white and black? Or did you somehow get it to use only one engine? Autogtp uses one engine to generate moves for both sides. I don't use sabaki so I don't know if it can do that. 

BTW I think we should somehow put the git HEAD hash into the autogtp and leelaz binaries, and are displayed in logs. I have burnt myself on this a few times already, git checkout a branch and forget to recompile both autogtp and leelaz.

**Comment by jkiliani** [comment]
@killerducky I recompiled leelaz in /leela-zero-next/src/ and copied the executable to /autogtp. With Sabaki, I connected two engines, both leelaz but configured by the command line arguments used in Autogtp (compete mode, currently only on /next).

**Comment by killerducky** [comment]
Oh I forgot this is for match mode, not self play. The autogtp -k and -d arguments don't save files for match mode.

**Comment by marcocalignano** [comment]
-k should work for match mode

**Comment by Ttl** [comment]
Self-play games are uploaded using `uploadData` which also handles copying the sgf file when -k option is enabled. Match games are uploaded with `uploadResult` which doesn't do the copying.

**Comment by killerducky** [comment]
Ok here is a much simpler way to get debug info. Run the bash shell (or translate to whatever shell you use) and try this. Modify path to networks as needed.

```
printf "auto" | ./leelaz -w ~/networks/920k.txt -p 10  -r 0  -m 0  -t 1  -d   --noponder  -s 0 > run1.txt 2>&1
printf "auto" | ./leelaz -w ~/networks/920k.txt -p 10  -r 0  -m 0  -t 1  -d   --noponder  -s 0 > run2.txt 2>&1

```
Then diff the files. There should only be differences in timing info like "### n/s" etc. I assume you will get some differences, then can you upload run1.txt and run2.txt?


BTW I'm using printf instead of echo because if you want multiple commands with \n it works better.

**Comment by marcocalignano** [comment]
@Ttl You are right I have already the fix in my client now I wait for the first match to see if it work if it does I make the pull request.

**Comment by gcp** [comment]
>Aren't there race conditions that introduce randomness existent in search anyway?

Not when running single-threaded.

We are sensitive to floating point rounding but same computer + same binary should be reproducible and in the configurations I tested it actaully is.
